### PR TITLE
afw review for tickets/HSC-1274: deblender optimization

### PR DESCRIPTION
**Pull request by TallJimbo**

### CODE REVIEW

File: include/lsst/afw/detection/Footprint.h [hunk 1]
```diff
@@ -237,12 +244,46 @@ void nearestFootprint(std::vector<Footprint::Ptr> const& foots,
...
+   This const version requires that both input footprints are
+   normalized (and will raise an exception if not).
+ */
+PTR(Footprint) mergeFootprints(Footprint const& foot1, Footprint const& foot2);
+
+/**
+   Merges two Footprints -- appends their peaks, and unions their
+   spans, returning a new Footprint.
+ */
+PTR(Footprint) mergeFootprints(Footprint& foot1, Footprint& foot2);
```
Comment (TallJimbo) [inline include/lsst/afw/detection/Footprint.h]: If this destroys the inputs, the documentation should say so.  If it doesn't, I don't understand why it exists separately from the overload above.

File: src/detection/Footprint.cc [hunk 2]
```diff
@@ -290,50 +290,39 @@ bool Footprint::contains(
...
+struct ClipSpansPredicate : public std::unary_function<PTR(Span) const&, bool> {
+    geom::Box2I const& bbox;
+    ClipSpansPredicate(geom::Box2I const& _bbox) : bbox(_bbox) {}
+    bool operator()(PTR(Span) const& spanPtr) const {
+        Span const& span = *spanPtr;
+        int const y = span.getY();
+        return (y < bbox.getMinY() || y > bbox.getMaxY() ||
+                span.getX0() > bbox.getMaxX() || span.getX1() < bbox.getMinX());
+    }
+};
```
Comment (TallJimbo) [inline src/detection/Footprint.cc]: A bit of vertical whitespace in this struct would make it much more readable.

Comment (TallJimbo) [inline src/detection/Footprint.cc]: Same for the one below (which may be my fault).

File: tests/footprint1.py [hunk 3]
```diff
@@ -36,6 +36,7 @@
 import math, sys
 import unittest
 import numpy
+import itertools
```
Comment (TallJimbo) [inline tests/footprint1.py]: I don't see this used below.

File: tests/footprint1.py [hunk 4]
```diff
@@ -950,6 +958,46 @@ def testClipToNonzero(self):
...
+        size = 15, 15
+        for span in ((3, 3, 9),
+                     (4, 2, 4),
+                     (4, 6, 7),
+                     (4, 9, 11),
+                     (5, 3, 9),
+                     (6, 6, 7),
+                     ):
+            self.foot.addSpanInSeries(*span)
+        self.foot.normalize()
```
Comment (TallJimbo) [inline tests/footprint1.py]: Given the complex logic involved in the implementation of `findEdgePixels`, I'd feel better with a more complex test case, maybe a Footprint derived from real data?

File: include/lsst/afw/detection/Footprint.h (hunk 1 above)
Comment (PaulPrice) [inline include/lsst/afw/detection/Footprint.h]: That bit's coming from the revert commit, so I think this is just propagating what's on the LSST side.  For that reason I'd prefer not to touch anything in this case unless it's important.

But I think the difference is solely that the non-const overload performs an explicit `normalize()` on each `Footprint`, and `normalize()` is not a `const` method.

**Pull request merged**